File: 0022-generate-parentheses/0022-generate-parentheses.java
class Solution {
    List<String> ans;

    private void backtrack(int n , int openCount , int closeCount , StringBuilder sb){
        if((sb.length() == 2 * n) && (closeCount == openCount)){
            ans.add(sb.toString());
            return;
        }

        if(openCount < n){
            sb.append("(");
            backtrack(n , openCount + 1 , closeCount , sb);
            sb.deleteCharAt(sb.length() - 1);
        }

        if(closeCount < openCount){
            sb.append(")");
            backtrack(n , openCount , closeCount + 1, sb);
            sb.deleteCharAt(sb.length() - 1);
        }
    }

    public List<String> generateParenthesis(int n) {
        ans = new ArrayList<>();
        backtrack(n , 0 , 0 , new StringBuilder());
        return ans;
    }
}

File: 0022-generate-parentheses/0022-generate-parentheses.py
class Solution:

    def __init__(self):
        self.ans = []

    def backtrack(self , n: int , open_count: int , close_count: int , path):
        if (len(path) == 2 * n) and (open_count == close_count):
            self.ans.append("".join(path))
            return

        if open_count < n:
            path.append('(')
            self.backtrack(n , open_count + 1 , close_count , path)
            path.pop()

        if close_count < open_count:
            path.append(')')
            self.backtrack(n , open_count , close_count + 1 , path)
            path.pop()


    def generateParenthesis(self, n: int) -> List[str]:
        self.backtrack(n , 0 , 0 , [])
        return self.ans